source/distributed/full_distributed_tria.cc

```diff
@@ -0,0 +1,508 @@
+          if (pp->first == pp->second)
+            continue;
+          AssertThrow(false, ExcMessage("Map has to be compressed!"));
+        }
+
+      std::map<unsigned int, std::set<unsigned int>> sets;
+      for (auto p : periodic_map)
+        sets[p.second] = std::set<unsigned int>();
+
+      for (auto p : periodic_map)
+        sets[p.second].insert(p.first);
+
+      std::map<unsigned int, std::set<unsigned int>> sets2;
+      for (auto &s : sets)
+        {
+          for (auto &ss : s.second)
+            sets2[ss] = s.second;
+        }
+
+      for (const auto &cell : this->active_cell_iterators())
+        if (cell->is_locally_owned())
+          {
+            for (unsigned int v = 0; v < GeometryInfo<dim>::vertices_per_cell;
+                 ++v)
+              vertex_of_own_cell[cell->vertex_index(v)] = true;
+          }
+
+      std::map<unsigned int, std::set<dealii::types::subdomain_id>> result;
+      for (const auto &cell : this->active_cell_iterators())
+        if (cell->is_ghost())
+          {
+            const types::subdomain_id owner = cell->subdomain_id();
+            for (unsigned int v = 0; v < GeometryInfo<dim>::vertices_per_cell;
+                 ++v)
+              {
+                if (vertex_of_own_cell[cell->vertex_index(v)])
+                  result[cell->vertex_index(v)].insert(owner);
+
+                // mark also nodes coinciding due to periodicity
+                auto t = sets2.find(cell->vertex_index(v));
+                if (t != sets2.end())
+                  for (auto i : t->second)
+                    if (vertex_of_own_cell[i])
+                      result[i].insert(owner);
+              }
+          }
+
+      return result;
+    }
+
+
+
+    template <int dim, int spacedim>
+    bool
+    Triangulation<dim, spacedim>::is_multilevel_hierarchy_constructed() const
+    {
+      return (settings & construct_multigrid_hierarchy);
+    }
+
+
+
+    template <int dim, int spacedim>
+    unsigned int
+    Triangulation<dim, spacedim>::coarse_cell_id_to_coarse_cell_index(
+      const types::coarse_cell_id coarse_cell_id) const
+    {
+      auto coarse_cell_index =
+        std::lower_bound(coarse_cell_id_to_coarse_cell_index_vector.begin(),
+                         coarse_cell_id_to_coarse_cell_index_vector.end(),
+                         coarse_cell_id,
+                         [](const auto &pair, const auto &val) {
+                           return pair.first < val;
+                         });
+      Assert(coarse_cell_index !=
+               coarse_cell_id_to_coarse_cell_index_vector.cend(),
+             ExcMessage("Coarse cell index not found!"));
+      return coarse_cell_index->second;
+    }
+
+
+
+    template <int dim, int spacedim>
+    types::coarse_cell_id
+    Triangulation<dim, spacedim>::coarse_cell_index_to_coarse_cell_id(
+      const unsigned int coarse_cell_index) const
+    {
+      return coarse_cell_index_to_coarse_cell_id_vector[coarse_cell_index];
```

Comment:
add an Assert for a valid coarse cell here. Maybe also check that the return value is not an invalid value (as I suggested setting for the one dummy cell).